### PR TITLE
Update thermal control design document

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
1. Update json configuration file with latest format
2. Add description for how to implement thermal policy
3. Update command line out put 
4. Add fan drawer and fan led management design

### CODE REVIEW

**Comment by Junchao-Mellanox** [comment]
@keboliu please help review.

File: thermal-control-design.md [hunk 1]
```diff
@@ -198,21 +396,53 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
+        raise NotImplementedError
+
+    def get_status_led(self, color):
+        raise NotImplementedError
+
+```
+
+The fan drawer object should follow:
+
+- If there are fans without drawer, we still add a drawer object for each of fan, and consider the drawer only has one fan.
```
Comment (keboliu) [inline thermal-control-design.md]: we need to handle this "virtual" fan drawer object carefully, we don't want to have this virtual fan drawer information to be showed in CLI.

File: thermal-control-design.md [hunk 2]
```diff
@@ -198,21 +396,53 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
+
+Existing platform API need changes:
 
+- Add two new member functions "get_num_fan_drawers" and "get_all_fan_drawers" to ChassisBase, vendor need to implement these two new functions to properly initialize chassis object with fan drawer objects.
 
-## 5. Potential ehhancement for Platform API
+## 6. Potential ehhancement for Platform API
 1. Why can't we propose different change events for different cpu/fan/optics?
 2. Verbose on API definition on threshold levels about Average/Max/Snapshot.
 3. Is there any API exposed for fanTray contain more than one fan?
```
Comment (keboliu) [inline thermal-control-design.md]: this line can be removed since now we are adding a fan drawer(fan tray)

File: thermal-control-design.md [hunk 3]
```diff
@@ -198,21 +396,53 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
+	admin@sonic# show platform fan
+	FAN    Speed      Direction  Presence     Status  Timestamp
+	-----  ---------  ---------  -----------  ------  -----------------
+	FAN 1  85%        Intake     Present      OK      20191112 09:38:16
+	FAN 2  60%        Exhaust    Not Present  Not OK  20191112 09:38:16
+
+
+## 5. Fan led management
+
+In current implementation, fan led management API is based on fan object. However, in many hardware platform, multiple fans are put in a fan drawer, they share one fan led. In such case, to control fan led in a fan object might cause problem. For example, one fan drawer has fan1 and fan2, they share a fan led, fan1 is broken and set led to red via fan1 object, then fan2 just recovers from a bad state and set led to green via fan2 object, the final fan led state will be green which is incorrect. To avoid confusing user, a new abstract class FanDrawerBase is added to sonic_platform_common and vendors should implement it. The FanDrawerBase class will look like:
```
Comment (keboliu) [inline thermal-control-design.md]: shall we extract fan drawer design as a separated section?

File: thermal-control-design.md [hunk 4]
```diff
@@ -198,21 +396,53 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
-	admin@sonic# show platform fanstatus
-	FAN    Speed      Direction      Timestamp
-	-----  ---------  ---------  -----------------
-	FAN 1  12919 RPM  Intake     20191112 09:38:16
-	FAN 2  13043 RPM  Exhaust    20191112 09:38:16
+	admin@sonic# show platform fan
+	FAN    Speed      Direction  Presence     Status  Timestamp
+	-----  ---------  ---------  -----------  ------  -----------------
+	FAN 1  85%        Intake     Present      OK      20191112 09:38:16
+	FAN 2  60%        Exhaust    Not Present  Not OK  20191112 09:38:16
```
Comment (keboliu) [inline thermal-control-design.md]: we shall update the show command output to reflect fan drawer info.  and also DB schema need to be changed to have fan drawer info, right?

**Review by keboliu — CHANGES_REQUESTED** [review]

File: thermal-control-design.md [hunk 5]
```diff
@@ -198,24 +397,67 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
-	admin@sonic# show platform fanstatus
-	FAN    Speed      Direction      Timestamp
-	-----  ---------  ---------  -----------------
-	FAN 1  12919 RPM  Intake     20191112 09:38:16
-	FAN 2  13043 RPM  Exhaust    20191112 09:38:16
+	admin@sonic# show platform fan
+	Drawer     FAN    Speed      Direction  Presence     Status  Timestamp
+	--------   -----  ---------  ---------  -----------  ------  -----------------
+	Drawer 1   FAN 1  85%        intake     Present      OK      20191112 09:38:16
+	           FAN 2  60%        intake     Present      OK      20191112 09:38:16
```
Comment (keboliu) [inline thermal-control-design.md]: suggest to add Drawer name to each of it's fan to avoid misunderstanding.

File: thermal-control-design.md [hunk 6]
```diff
@@ -198,24 +397,67 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
+	Drawer     FAN    Speed      Direction  Presence     Status  Timestamp
+	--------   -----  ---------  ---------  -----------  ------  -----------------
+	N/A        FAN 1  85%        intake     Present      OK      20191112 09:38:16
+	N/A        FAN 2  60%        intake     Present      OK      20191112 09:38:16
+    N/A        FAN 3  75%        exhaust    Present      Not OK  20191112 09:38:16
+    N/A        FAN 4  65%        exhaust    Present      Not OK  20191112 09:38:16 
+
+## 5. Fan drawer implementation
+
+In current implementation, fan led management API is based on fan object. However, in many hardware platform, multiple fans are put in a fan drawer, they share one fan led. In such case, to control fan led in a fan object might cause problem. For example, one fan drawer has fan1 and fan2, they share a fan led, fan1 is broken and set led to red via fan1 object, then fan2 just recovers from a bad state and set led to green via fan2 object, the final fan led state will be green which is incorrect. To avoid confusing user, a new abstract class FanDrawerBase is added to sonic_platform_common and vendors should implement it. The FanDrawerBase class will look like:
```
Comment (keboliu) [inline thermal-control-design.md]: I think adding fan drawer is not only for led,  it also gives information about real HW layout, which fan are in which drawer.

File: thermal-control-design.md [hunk 7]
```diff
@@ -198,24 +397,67 @@ We don't have a CLI for fan status getting yet, new CLI for fan status could be
...
+    def get_status_led(self, color):
+        raise NotImplementedError
+
+    def is_virtual_drawer(self):
+        raise NotImplementedError
+```
+
+The fan drawer object should follow:
+
+- If there are fans without drawer, we still add a virtual drawer object for each of fan, and consider the drawer only has one fan. We don't show virtual drawer in CLI.
```
Comment (keboliu) [inline thermal-control-design.md]: should make it clear that thermalctld go through drawer first, so have to have a virtual drawer object even no drawer.

**Review by keboliu — APPROVED** [review]

**Pull request closed without merging**